### PR TITLE
Add drag and drop functionality to the recipient views

### PR DESCRIPTION
**Pull request by rhari991**

### CODE REVIEW

File: k9mail/src/main/java/com/fsck/k9/view/RecipientSelectView.java [hunk 1]
```diff
@@ -100,7 +105,90 @@ private void initView(Context context) {
...
 
-        setLongClickable(true);
+        addDragAndDropFunctionality(context);
+    }
+
+    public void setDragListener(DragListener listener) {
+        dragListener = listener;
+    }
+
+    private void addDragAndDropFunctionality(Context context) {
```
Comment (Valodim) [inline k9mail/src/main/java/com/fsck/k9/view/RecipientSelectView.java]: this functionality deserves a class of its own, something like `RecipientDragAndDropUseCase`, that on creation receives the state it needs and a reference to the MvpView to do UI things, and exists as long as a drag is going on.

File: k9mail/src/main/java/com/fsck/k9/activity/MessageCompose.java [hunk 2]
```diff
@@ -1797,4 +1808,13 @@ public int getTitleResource() {
             return titleResource;
         }
     }
+
+    private class RecipientDropListener implements View.OnDragListener {
```
Comment (Valodim) [inline k9mail/src/main/java/com/fsck/k9/activity/MessageCompose.java]: should be part of `DragAndDropUseCase` (see below)

File: k9mail/src/main/java/com/fsck/k9/view/RecipientSelectView.java (hunk 1 above)
Comment (rhari991) [inline k9mail/src/main/java/com/fsck/k9/view/RecipientSelectView.java]: Do you mean a reference to the RecipientSelectView ? There is very little interaction between RecipientMvpView and RecipientSelectView for drag and drop (only one call to onClickRecipientExpander()).

**Comment by rhari991** [comment]
I made the chages in PR #2511

File: k9mail/src/main/java/com/fsck/k9/view/RecipientSelectView.java (hunk 1 above)
Comment (rhari991) [inline k9mail/src/main/java/com/fsck/k9/view/RecipientSelectView.java]: I've implemented this, but I'm sure if I've done it correctly.

File: k9mail/src/main/java/com/fsck/k9/view/RecipientSelectView.java [hunk 3]
```diff
@@ -121,7 +209,7 @@ private View inflateLayout() {
         return layoutInflater.inflate(R.layout.recipient_token_item, null, false);
     }
 
-    private void bindObjectView(Recipient recipient, View view) {
+    private void bindObjectView(final Recipient recipient, View view) {
```
Comment (rhari991) [inline k9mail/src/main/java/com/fsck/k9/view/RecipientSelectView.java]: Reverted, sorrry.

File: k9mail/src/main/java/com/fsck/k9/activity/compose/RecipientMvpView.java [hunk 4]
```diff
@@ -54,6 +54,10 @@
     private final ViewAnimator recipientExpanderContainer;
     private final ViewAnimator cryptoSpecialModeIndicator;
     private RecipientPresenter presenter;
+    private boolean dragTakingPlace = false;
```
Comment (rhari991) [inline k9mail/src/main/java/com/fsck/k9/activity/compose/RecipientMvpView.java]: I've moved all of the state code to the presenter.

File: k9mail/src/main/java/com/fsck/k9/activity/MessageCompose.java (hunk 2 above)
Comment (rhari991) [inline k9mail/src/main/java/com/fsck/k9/activity/MessageCompose.java]: I'm not sure how to do this.

**Comment by rhari991** [comment]
Sorry for the long wait ! College work has been keeping me busy.

**Comment by cketti** [comment]
Loving the feature. But I'm hoping we can replace the tokenautocomplete library with something better.

**Pull request closed without merging**